File: legacy/packages/lexer/src/raw-lexer.ts
import { KEYWORDS, type Token, type TokenKind } from "./tokens";

export class LexError extends Error {
  constructor(
    message: string,
    public line: number,
    public column: number,
  ) {
    super(`${message} at ${line}:${column}`);
    this.name = "LexError";
  }
}

interface RawToken extends Token {
  hadSpaceBefore: boolean;
  hadSpaceAfter: boolean;
}

export function rawLex(source: string): Token[] {
  const tokens: RawToken[] = [];
  let i = 0;
  let line = 1;
  let column = 1;

  const peek = (n = 0) => source[i + n] ?? "";
  const advance = () => {
    const ch = source[i++];
    if (ch === "\n") {
      line++;
      column = 1;
    } else {
      column++;
    }
    return ch;
  };

  const makeToken = (
    kind: TokenKind,
    value: string | undefined,
    startLine: number,
    startCol: number,
    startOffset: number,
    hadSpaceBefore: boolean,
    hadSpaceAfter: boolean,
  ): RawToken => ({
    kind,
    value,
    line: startLine,
    column: startCol,
    offset: startOffset,
    hadSpaceBefore,
    hadSpaceAfter,
  });

  const skipComment = () => {
    advance(); // second -
    while (i < source.length && peek() !== "\n") advance();
  };

  const readString = (startLine: number, startCol: number, startOffset: number, hadSpaceBefore: boolean) => {
    const quote = advance();
    if (quote === '"') {
      if (peek() === '"' && peek(1) === '"') {
        advance();
        advance();
        let value = "";
        while (i < source.length) {
          if (peek() === '"' && peek(1) === '"' && peek(2) === '"') {
            advance();
            advance();
            advance();
            break;
          }
          value += advance();
        }
        const hadSpaceAfter = peek() === " " || peek() === "\t";
        tokens.push(makeToken("STRING", value, startLine, startCol, startOffset, hadSpaceBefore, hadSpaceAfter));
        return;
      }
      let value = "";
      while (i < source.length && peek() !== '"') {
        if (peek() === "\\") {
          advance();
          value += advance();
        } else {
          value += advance();
        }
      }
      if (peek() !== '"') throw new LexError("Unterminated string", line, column);
      advance();
      const hadSpaceAfter = peek() === " " || peek() === "\t";
      tokens.push(makeToken("STRING", value, startLine, startCol, startOffset, hadSpaceBefore, hadSpaceAfter));
      return;
    }
  };

  const readNumber = (startLine: number, startCol: number, startOffset: number, hadSpaceBefore: boolean) => {
    let num = "";
    while (/[0-9]/.test(peek())) num += advance();
    if (peek() === "." && /[0-9]/.test(peek(1))) {
      num += advance();
      while (/[0-9]/.test(peek())) num += advance();
      const hadSpaceAfter = peek() === " " || peek() === "\t";
      tokens.push(makeToken("FLOAT", num, startLine, startCol, startOffset, hadSpaceBefore, hadSpaceAfter));
      return;
    }
    const hadSpaceAfter = peek() === " " || peek() === "\t";
    tokens.push(makeToken("INT", num, startLine, startCol, startOffset, hadSpaceBefore, hadSpaceAfter));
  };

  const MODIFIER_KEYWORDS = new Set(["constraint", "default", "optional", "fetched"]);

  const readIdent = (startLine: number, startCol: number, startOffset: number, hadSpaceBefore: boolean) => {
    let ident = "";
    while (/[a-zA-Z0-9_\-]/.test(peek())) ident += advance();
    const hadSpaceAfter = peek() === " " || peek() === "\t";
    const kw = KEYWORDS[ident];
    const isConfigKey = peek() === ":" && peek(1) !== ":";
    if (kw && !(isConfigKey && !MODIFIER_KEYWORDS.has(ident))) {
      tokens.push(makeToken(kw, ident, startLine, startCol, startOffset, hadSpaceBefore, hadSpaceAfter));
    } else {
      tokens.push(makeToken("IDENT", ident, startLine, startCol, startOffset, hadSpaceBefore, hadSpaceAfter));
    }
  };

  while (i < source.length) {
    const ch = peek();

    if (ch === " " || ch === "\t" || ch === "\r") {
      advance();
      continue;
    }

    if (ch === "\n") {
      const startLine = line;
      const startCol = column;
      const startOffset = i;
      advance();
      tokens.push(makeToken("NEWLINE", undefined, startLine, startCol, startOffset, false, false));
      continue;
    }

    const hadSpaceBefore = tokens.length > 0 && tokens[tokens.length - 1]!.hadSpaceAfter;
    const startLine = line;
    const startCol = column;
    const startOffset = i;

    if (ch === "-" && peek(1) === "-") {
      skipComment();
      continue;
    }

    if (ch === '"') {
      readString(startLine, startCol, startOffset, hadSpaceBefore);
      continue;
    }

    if (/[0-9]/.test(ch)) {
      readNumber(startLine, startCol, startOffset, hadSpaceBefore);
      continue;
    }

    if (/[a-zA-Z_]/.test(ch)) {
      readIdent(startLine, startCol, startOffset, hadSpaceBefore);
      continue;
    }

    const two = ch + peek(1);
    const three = two + peek(2);

    const pushSimple = (kind: TokenKind, len: number, value?: string) => {
      for (let j = 0; j < len; j++) advance();
      const hadSpaceAfter = peek() === " " || peek() === "\t";
      tokens.push(makeToken(kind, value, startLine, startCol, startOffset, hadSpaceBefore, hadSpaceAfter));
    };

    if (three === "<-=") throw new LexError("Unexpected '<-='", startLine, startCol);
    if (two === "<-") { pushSimple("BIND", 2); continue; }
    if (three === ">>=") { pushSimple("SEQ", 3, ">>="); continue; }
    if (two === ">>") { pushSimple("SEQ", 2, ">>"); continue; }
    if (two === "==") { pushSimple("EQ", 2); continue; }
    if (two === "/=") { pushSimple("NEQ", 2); continue; }
    if (two === "<=") { pushSimple("LE", 2); continue; }
    if (two === ">=") { pushSimple("GE", 2); continue; }
    if (two === "->") { pushSimple("ARROW", 2); continue; }
    if (two === "&&") { pushSimple("AND", 2); continue; }
    if (two === "||") { pushSimple("OR", 2); continue; }

    if (ch === ".") {
      const spaceAfter = peek(1) === " " || peek(1) === "\t";
      if (hadSpaceBefore && spaceAfter) {
        pushSimple("COMPOSE_DOT", 1);
      } else {
        pushSimple("FIELD_DOT", 1);
      }
      continue;
    }

    if (ch === "-") {
      if (hadSpaceBefore || peek(1) === " " || peek(1) === "\t" || !/[0-9a-zA-Z_]/.test(peek(1))) {
        pushSimple("MINUS", 1);
      } else {
        readIdent(startLine, startCol, startOffset, hadSpaceBefore);
      }
      continue;
    }

    switch (ch) {
      case "{": pushSimple("LBRACE", 1); break;
      case "}": pushSimple("RBRACE", 1); break;
      case ";": pushSimple("SEMI", 1); break;
      case "(": pushSimple("LPAREN", 1); break;
      case ")": pushSimple("RPAREN", 1); break;
      case "[": pushSimple("LBRACKET", 1); break;
      case "]": pushSimple("RBRACKET", 1); break;
      case ",": pushSimple("COMMA", 1); break;
      case ":": pushSimple("COLON", 1); break;
      case "=": pushSimple("EQUALS", 1); break;
      case "|": pushSimple("PIPE", 1); break;
      case "\\": pushSimple("BACKSLASH", 1); break;
      case "$": pushSimple("DOLLAR", 1); break;
      case "+": pushSimple("PLUS", 1); break;
      case "*": pushSimple("STAR", 1); break;
      case "/": pushSimple("SLASH", 1); break;
      case "<": pushSimple("LT", 1); break;
      case ">": pushSimple("GT", 1); break;
      default:
        throw new LexError(`Unexpected character '${ch}'`, startLine, startCol);
    }
  }

  tokens.push(makeToken("EOF", undefined, line, column, i, false, false));
  return tokens.map(({ hadSpaceBefore: _b, hadSpaceAfter: _a, ...t }) => t);
}

File: Sources/MoonLexer/RawLexer.swift
import Foundation

public struct LexError: Error, CustomStringConvertible {
    public var message: String
    public var line: Int
    public var column: Int

    public init(_ message: String, line: Int, column: Int) {
        self.message = message
        self.line = line
        self.column = column
    }

    public var description: String {
        "\(message) at \(line):\(column)"
    }
}

private struct RawToken {
    var token: Token
    var hadSpaceBefore: Bool
    var hadSpaceAfter: Bool
}

private func isAsciiLetter(_ s: UnicodeScalar) -> Bool {
    (65...90).contains(s.value) || (97...122).contains(s.value)
}

private func isAsciiDigit(_ s: UnicodeScalar) -> Bool {
    (48...57).contains(s.value)
}

public func rawLex(_ source: String) throws -> [Token] {
    let scalars = Array(source.unicodeScalars)
    var tokens: [RawToken] = []
    var i = 0
    var line = 1
    var column = 1

    func peek(_ n: Int = 0) -> UnicodeScalar? {
        let idx = i + n
        guard idx < scalars.count else { return nil }
        return scalars[idx]
    }

    @discardableResult
    func advance() -> UnicodeScalar {
        let ch = scalars[i]
        i += 1
        if ch == "\n" {
            line += 1
            column = 1
        } else {
            column += 1
        }
        return ch
    }

    func makeToken(
        kind: TokenKind,
        value: String?,
        startLine: Int,
        startCol: Int,
        startOffset: Int,
        hadSpaceBefore: Bool,
        hadSpaceAfter: Bool
    ) -> RawToken {
        RawToken(
            token: Token(kind: kind, value: value, line: startLine, column: startCol, offset: startOffset),
            hadSpaceBefore: hadSpaceBefore,
            hadSpaceAfter: hadSpaceAfter
        )
    }

    func hasSpaceAfter() -> Bool {
        peek() == " " || peek() == "\t"
    }

    func skipComment() {
        _ = advance()
        while i < scalars.count {
            let ch = peek()!
            if ch == "\n" || ch == "\r" { break }
            _ = advance()
        }
    }

    func readString(startLine: Int, startCol: Int, startOffset: Int, hadSpaceBefore: Bool) throws {
        _ = advance()
        if peek() == "\"", peek(1) == "\"" {
            _ = advance()
            _ = advance()
            var value = ""
            while i < scalars.count {
                if peek() == "\"", peek(1) == "\"", peek(2) == "\"" {
                    _ = advance()
                    _ = advance()
                    _ = advance()
                    break
                }
                value.unicodeScalars.append(advance())
            }
            tokens.append(makeToken(
                kind: .string, value: value,
                startLine: startLine, startCol: startCol, startOffset: startOffset,
                hadSpaceBefore: hadSpaceBefore, hadSpaceAfter: hasSpaceAfter()
            ))
            return
        }

        var value = ""
        while i < scalars.count, peek() != "\"" {
            if peek() == "\\" {
                _ = advance()
                value.unicodeScalars.append(advance())
            } else {
                value.unicodeScalars.append(advance())
            }
        }
        guard peek() == "\"" else {
            throw LexError("Unterminated string", line: line, column: column)
        }
        _ = advance()
        tokens.append(makeToken(
            kind: .string, value: value,
            startLine: startLine, startCol: startCol, startOffset: startOffset,
            hadSpaceBefore: hadSpaceBefore, hadSpaceAfter: hasSpaceAfter()
        ))
    }

    func readNumber(startLine: Int, startCol: Int, startOffset: Int, hadSpaceBefore: Bool) {
        var num = ""
        while let ch = peek(), isAsciiDigit(ch) {
            num.unicodeScalars.append(advance())
        }
        if peek() == ".", let next = peek(1), isAsciiDigit(next) {
            num.unicodeScalars.append(advance())
            while let ch = peek(), isAsciiDigit(ch) {
                num.unicodeScalars.append(advance())
            }
            tokens.append(makeToken(
                kind: .float, value: num,
                startLine: startLine, startCol: startCol, startOffset: startOffset,
                hadSpaceBefore: hadSpaceBefore, hadSpaceAfter: hasSpaceAfter()
            ))
            return
        }
        tokens.append(makeToken(
            kind: .int, value: num,
            startLine: startLine, startCol: startCol, startOffset: startOffset,
            hadSpaceBefore: hadSpaceBefore, hadSpaceAfter: hasSpaceAfter()
        ))
    }

    func readIdent(startLine: Int, startCol: Int, startOffset: Int, hadSpaceBefore: Bool) {
        var ident = ""
        while let ch = peek(), isAsciiLetter(ch) || isAsciiDigit(ch) || ch == "_" || ch == "-" {
            ident.unicodeScalars.append(advance())
        }
        let spaceAfter = hasSpaceAfter()
        let peekChar: (Int) -> Character = { n in
            guard let s = peek(n) else { return "\0" }
            return Character(s)
        }
        if let kw = keywordKind(for: ident, peek: peekChar, hadSpaceAfter: spaceAfter) {
            tokens.append(makeToken(
                kind: kw, value: ident,
                startLine: startLine, startCol: startCol, startOffset: startOffset,
                hadSpaceBefore: hadSpaceBefore, hadSpaceAfter: spaceAfter
            ))
        } else {
            tokens.append(makeToken(
                kind: .ident, value: ident,
                startLine: startLine, startCol: startCol, startOffset: startOffset,
                hadSpaceBefore: hadSpaceBefore, hadSpaceAfter: spaceAfter
            ))
        }
    }

    func pushSimple(kind: TokenKind, len: Int, value: String?, startLine: Int, startCol: Int, startOffset: Int, hadSpaceBefore: Bool) {
        for _ in 0..<len { _ = advance() }
        tokens.append(makeToken(
            kind: kind, value: value,
            startLine: startLine, startCol: startCol, startOffset: startOffset,
            hadSpaceBefore: hadSpaceBefore, hadSpaceAfter: hasSpaceAfter()
        ))
    }

    func emitNewline(startLine: Int, startCol: Int, startOffset: Int) {
        if peek() == "\r" { _ = advance() }
        if peek() == "\n" { _ = advance() }
        tokens.append(makeToken(
            kind: .newline, value: nil,
            startLine: startLine, startCol: startCol, startOffset: startOffset,
            hadSpaceBefore: false, hadSpaceAfter: false
        ))
    }

    while i < scalars.count {
        guard let ch = peek() else { break }

        if ch == " " || ch == "\t" {
            _ = advance()
            continue
        }

        if ch == "\r" || ch == "\n" {
            let startLine = line
            let startCol = column
            let startOffset = i
            emitNewline(startLine: startLine, startCol: startCol, startOffset: startOffset)
            continue
        }

        let hadSpaceBefore = tokens.last?.hadSpaceAfter ?? false
        let startLine = line
        let startCol = column
        let startOffset = i

        if ch == "-", peek(1) == "-" {
            skipComment()
            continue
        }

        if ch == "\"" {
            try readString(startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
            continue
        }

        if isAsciiDigit(ch) {
            readNumber(startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
            continue
        }

        if isAsciiLetter(ch) || ch == "_" {
            readIdent(startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
            continue
        }

        let two: String = {
            guard let a = peek(), let b = peek(1) else { return String(ch) }
            return String(Character(a)) + String(Character(b))
        }()
        let three: String = {
            guard let a = peek(), let b = peek(1), let c = peek(2) else { return two }
            return two + String(Character(c))
        }()

        if three == "<-=" {
            throw LexError("Unexpected '<-='", line: startLine, column: startCol)
        }
        if two == "<-" { pushSimple(kind: .bind, len: 2, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore); continue }
        if three == ">>=" { pushSimple(kind: .seq, len: 3, value: ">>=", startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore); continue }
        if two == ">>" { pushSimple(kind: .seq, len: 2, value: ">>", startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore); continue }
        if two == "==" { pushSimple(kind: .eq, len: 2, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore); continue }
        if two == "/=" { pushSimple(kind: .neq, len: 2, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore); continue }
        if two == "<=" { pushSimple(kind: .le, len: 2, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore); continue }
        if two == ">=" { pushSimple(kind: .ge, len: 2, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore); continue }
        if two == "->" { pushSimple(kind: .arrow, len: 2, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore); continue }
        if two == "&&" { pushSimple(kind: .and, len: 2, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore); continue }
        if two == "||" { pushSimple(kind: .or, len: 2, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore); continue }

        if ch == "." {
            let spaceAfter = peek(1) == " " || peek(1) == "\t"
            if hadSpaceBefore && spaceAfter {
                pushSimple(kind: .composeDot, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
            } else {
                pushSimple(kind: .fieldDot, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
            }
            continue
        }

        if ch == "-" {
            let next = peek(1)
            if hadSpaceBefore || next == " " || next == "\t" || next == nil
                || !(next.map { isAsciiLetter($0) || isAsciiDigit($0) || $0 == "_" } ?? false) {
                pushSimple(kind: .minus, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
            } else {
                readIdent(startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
            }
            continue
        }

        switch ch {
        case "{": pushSimple(kind: .lbrace, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "}": pushSimple(kind: .rbrace, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case ";": pushSimple(kind: .semi, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "(": pushSimple(kind: .lparen, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case ")": pushSimple(kind: .rparen, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "[": pushSimple(kind: .lbracket, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "]": pushSimple(kind: .rbracket, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case ",": pushSimple(kind: .comma, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case ":": pushSimple(kind: .colon, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "=": pushSimple(kind: .equals, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "|": pushSimple(kind: .pipe, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "\\": pushSimple(kind: .backslash, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "$": pushSimple(kind: .dollar, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "+": pushSimple(kind: .plus, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "*": pushSimple(kind: .star, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "/": pushSimple(kind: .slash, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case "<": pushSimple(kind: .lt, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        case ">": pushSimple(kind: .gt, len: 1, value: nil, startLine: startLine, startCol: startCol, startOffset: startOffset, hadSpaceBefore: hadSpaceBefore)
        default:
            throw LexError("Unexpected character '\(Character(ch))'", line: startLine, column: startCol)
        }
    }

    tokens.append(makeToken(
        kind: .eof, value: nil,
        startLine: line, startCol: column, startOffset: i,
        hadSpaceBefore: false, hadSpaceAfter: false
    ))
    return tokens.map(\.token)
}